clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp

```diff
@@ -0,0 +1,253 @@
+                on(Member)),
+            callExpr(argumentCountIs(1), hasArgument(0, Member),
+                     hasDeclaration(functionDecl(
+                         IsBegin ? hasAnyName("::std::begin", "::std::cbegin")
+                                 : hasAnyName("::std::end", "::std::cend"))))));
+}
+static ast_matchers::internal::Matcher<CallExpr>
+makeMatcherPair(StringRef State,
+                const UseRangesCheck::Replacer::Indexes &Indexes) {
+  auto ArgPostfix = std::to_string(Indexes.BeginArg);
+  SmallString<64> ID = {BoundCall, State};
+  return callExpr(argumentCountAtLeast(
+                      std::max(Indexes.BeginArg, Indexes.EndArg) + 1),
+                  hasArgument(Indexes.BeginArg, makeMatcher(true, ArgPostfix)),
+                  hasArgument(Indexes.EndArg, makeMatcher(false, ArgPostfix)))
+      .bind(ID);
+}
+
+void UseRangesCheck::registerMatchers(MatchFinder *Finder) {
+  Replaces = GetReplacerMap();
+  llvm::DenseSet<ArrayRef<ArrayRef<Replacer::Indexes>>> Seen;
+  for (auto I = Replaces.begin(), E = Replaces.end(); I != E; ++I) {
+    const auto &Replacer = I->getValue();
+    const auto &Signatures = Replacer->getReplacementSignatures();
+    if (Seen.contains(Signatures))
+      continue;
+    assert(!Signatures.empty() &&
+           llvm::all_of(Signatures, [](auto index) { return !index.empty(); }));
+    std::vector<StringRef> Names(1, I->getKey());
+    for (auto J = std::next(I); J != E; ++J) {
+      if (J->getValue()->getReplacementSignatures() == Signatures) {
+        Names.push_back(J->getKey());
+      }
+    }
+    std::vector<ast_matchers::internal::DynTypedMatcher> TotalMatchers;
+    // As we match on the first matched signature, we need to ensure that any
+    SmallVector<ArrayRef<Replacer::Indexes>> SigVec(Signatures);
+    llvm::sort(SigVec, [](auto &L, auto &R) { return R.size() < L.size(); });
+    for (const auto &Signature : SigVec) {
+      std::vector<ast_matchers::internal::DynTypedMatcher> Matchers;
+      for (const auto &ArgPair : Signature) {
+        Matchers.push_back(makeMatcherPair(getFullPrefix(Signature), ArgPair));
+      }
+      TotalMatchers.push_back(
+          ast_matchers::internal::DynTypedMatcher::constructVariadic(
+              ast_matchers::internal::DynTypedMatcher::VO_AllOf,
+              ASTNodeKind::getFromNodeKind<CallExpr>(), std::move(Matchers)));
+    }
+    Finder->addMatcher(
+        callExpr(
+            callee(functionDecl(hasAnyName(std::move(Names))).bind(FuncDecl)),
+            ast_matchers::internal::DynTypedMatcher::constructVariadic(
+                ast_matchers::internal::DynTypedMatcher::VO_AnyOf,
+                ASTNodeKind::getFromNodeKind<CallExpr>(),
+                std::move(TotalMatchers))
+                .convertTo<CallExpr>()),
+        this);
+  }
+}
+
+// Returns the proper token based end location of \p E.
+static SourceLocation exprLocEnd(const Expr *E, const ASTContext &Context) {
```

Comment:
Consider moving it to LexerUtils

Reply:
I just removed this call, had one call site and just inlined instead